tests/integration/security/util/reachability/context.go

```diff
@@ -171,69 +113,104 @@ func (rc *Context) Run(testCases []TestCase) {
+						// only hit same cluster multiversion services
+						apps.Multiversion.Match(echo.InCluster(client.Config().Cluster)),
+						// only hit same cluster naked services
+						apps.Naked.Match(echo.InCluster(client.Config().Cluster)),
+						// only hit same cluster vm services
+						apps.VM.Match(echo.InCluster(client.Config().Cluster)),
+						// only hit same cluster headless services
+						apps.HeadlessNaked.Match(echo.InCluster(client.Config().Cluster)),
 					}
-					for _, opts := range *copts {
-						// Copy the loop variables so they won't change for the subtests.
-						src := src
-						dest := dest
-						opts := opts
-						onPreRun := c.OnRun
-
-						// Set the target on the call options.
-						opts.Target = dest
-
-						if c.Include(src, opts) {
-							expectSuccess := c.ExpectSuccess(src, opts)
-
-							subTestName := fmt.Sprintf("%s->%s://%s:%s%s",
-								src.Config().Service,
-								opts.Scheme,
-								dest.Config().Service,
-								opts.PortName,
-								opts.Path)
-
-							ctx.NewSubTest(subTestName).
-								RunParallel(func(ctx framework.TestContext) {
-									if onPreRun != nil {
-										onPreRun(ctx, src, opts)
-									}
-
-									checker := connection.Checker{
-										From:          src,
-										Options:       opts,
-										ExpectSuccess: expectSuccess,
-									}
-									checker.CheckOrFail(ctx)
-								})
+
+					for _, destinations := range destinationSets {
+						client := client
+						destinations := destinations
+						destClusters := destinations.Clusters()
+						if len(destClusters) == 0 {
+							continue
+						}
+						// grabbing the 0th assumes all echos in destinations have the same service name
+						destination := destinations[0]
+						if (apps.IsHeadless(client) || apps.IsHeadless(destination) || apps.IsNaked(client)) && len(destClusters) > 1 {
+							// TODO(landow) fix DNS issues with multicluster/VMs/headless
+							continue
+						}
+						callCount := 1
+						if len(destClusters) > 1 {
+							// so we can validate all clusters are hit
+							callCount = util.CallsPerCluster * len(destClusters)
 						}
+
+						ctx.NewSubTest(fmt.Sprintf("%s->%s",
+							client.Config().Service, destination.Config().Service)).Run(func(ctx framework.TestContext) {
+							copts := &callOptions
+							// If test case specified service call options, use that instead.
+							if c.CallOpts != nil {
+								copts = &c.CallOpts
+							}
+							for _, opts := range *copts {
+								// Copy the loop variables so they won't change for the subtests.
+								src := client
+								dest := destination
+								opts := opts
+								onPreRun := c.OnRun
+
+								// Set the target on the call options.
+								opts.Target = dest
+								opts.Count = callCount
+
+								if c.Include(src, opts) {
+									expectSuccess := c.ExpectSuccess(src, opts)
+
+									subTestName := fmt.Sprintf("%s->%s in cluster %s://%s:%s%s",
```

Comment:
Probably good to remove the service names from this. The subtest names `security: TestReachability/beta-mtls-on/a->a/a->ws_in_cluster_cluster-0://a:http` look confusing, it should be formatted more like 

`TestReachability/beta-mtls-on/a/in_cluster_0/to_b/ws`